### PR TITLE
[PyTorch Edge][type] Add type support for NamedTuple custom class (export)

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62612

Add type support for namedtule custom class. For the namedtuple type, it will deserailize to the following format in string
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```

If it's nested, it will be
```
"__torch__.A[
    NamedTuple, [
        [field_name_a, __torch__.B [
            NamedTuple, [
                [field_name_b, __torch__.C [
                    NamedTuple, [
                      [field_name_c_1, Tensor],
                      [field_name_c_2, Tuple[Tensor, Tensor]],
                    ]
                ]
                ]
            ]
        ]
        ]
    ]
]
"
```
The nametuple type includes both `collection` and `typing`.
```

from typing import NamedTuple
from collections import namedtuple
```

It will be a forward incompatible change. However this type is never supported and exported before and we don't have a proper way to backport it. The optimum solution to ship this change is probably
1. Update the change for import without the change to export. So the runtime can read the new format, but no new format will be exported.
2. Update the change to export the new type. So runtime can export new format.

For the following example:
```
class Foo(NamedTuple):
    id: torch.Tensor

class Bar(torch.nn.Module):
    def __init__(self):
        super(Bar, self).__init__()
        self.foo = Foo(torch.tensor(1))

    def forward(self, a: torch.Tensor):
        self.foo = Foo(a)
        return self.foo
```
The new bytecode.pkl will be
```
(6,
 ('__torch__.mobile.test_lite_script_type.MyTestModule.forward',
  (('instructions',
    (('STOREN', 1, 2),
     ('DROPR', 1, 0),
     ('MOVE', 2, 0),
     ('LIST_CONSTRUCT', 0, 1),
     ('NAMED_TUPLE_CONSTRUCT', 1, 1),
     ('RET', 0, 0))),
   ('operators', ()),
   ('constants', ()),
   ('types',
    ('List[Tensor]',
     '__torch__.mobile.test_lite_script_type.myNamedTuple[NamedTuple, [[a, '
     'List[Tensor]]]]')),
   ('register_size', 2)),
  (('arguments',
    ((('name', 'self'),
      ('type', '__torch__.mobile.test_lite_script_type.MyTestModule'),
      ('default_value', None)),
     (('name', 'a'), ('type', 'Tensor'), ('default_value', None)))),
   ('returns',
    ((('name', ''),
      ('type', '__torch__.mobile.test_lite_script_type.myNamedTuple'),
      ('default_value', None)),)))))
```

Differential Revision: [D30004716](https://our.internmc.facebook.com/intern/diff/D30004716/)

Differential Revision: [D30004716](https://our.internmc.facebook.com/intern/diff/D30004716)

### CODE REVIEW

File: torch/csrc/jit/mobile/import.cpp [hunk 1]
```diff
@@ -424,17 +424,53 @@ void BytecodeDeserializer::parseMethods(
...
+          TORCH_CHECK(
+              classType,
+              "The implementation of class ",
+              qn.qualifiedName(),
+              " cannot be found.");
+          function->append_type(classType);
+        } else {
+          function->append_type(c10::parseType(t.toStringRef()));
+        }
       } else {
```
Comment (JacobSzwejbka) [inline torch/csrc/jit/mobile/import.cpp:440]: what type is t in this else?

File: torch/csrc/jit/mobile/import.cpp [hunk 2]
```diff
@@ -424,17 +424,53 @@ void BytecodeDeserializer::parseMethods(
...
+             name_type_pairs[1].toTuple()->elements()) {
+          std::vector<IValue> name_type_vector =
+              name_type_pair.toTuple()->elements();
+          std::string field_name = name_type_vector[0].toString()->string();
+          TypePtr field_type =
+              c10::parseType(name_type_vector[1].toString()->string());
+          field_names.emplace_back(field_name);
+          field_types.emplace_back(field_type);
+        }
+        // Create the NamedTuple type after reading from the tuple, and add it to function
```
Comment (JacobSzwejbka) [inline torch/csrc/jit/mobile/import.cpp:470]: And then this is me just not understanding how mobile deserialization works, but its not intuitive to me that we are adding the NamedTuple type fo the function its called from instead of say the module as a member or something. Can you maybe elaborate on this?

**Comment by raziel** [comment]
> However this type is never supported and exported before and we don't have a proper way to backport it.

The way I think about it, this is not something meant to be backported since as you said we never had a representation for it.


> The optimum solution to ship this change is probably:
> Update the change for import without the change to export. So the runtime can read the new format, but no new format will be exported.
> Update the change to export the new type. So runtime can export new format.

I think what you mean by "backporting" is that this change is not tied to a bytecode version change, again because before we had no equivalent representation (basically we could not express programs with the newly added semantics).
With that in mind, it makes sense there's no version tied to this change (and hence no version bump, and no backport).

Now, what you mention about just staging support for runtime and then for producing programs with this new functionality makes sense (but again, it's separate from the "versioning" discussion).


> It will be a forward compatibility change.

Regarding checking for compatibility between a program and the runtime.
Per your comment, the change is FC (i.e. an old runtime can read models with NamedTuple). I assume it is BC as well (an old model, which couldn't have the feature, would naturally be able to run), right?

File: test/cpp/jit/test_lite_trainer.cpp [hunk 3]
```diff
@@ -18,6 +18,61 @@
 namespace torch {
 namespace jit {
 
+TEST(LiteTrainerTest, CustomOp) {
+  // std::string mobile_model =
```
Comment (raziel) [inline test/cpp/jit/test_lite_trainer.cpp:22]: Maybe clean this up if this is ready for review?

File: torch/csrc/jit/mobile/import.cpp [hunk 4]
```diff
@@ -424,17 +424,53 @@ void BytecodeDeserializer::parseMethods(
 
     static const c10::QualifiedName classPrefix = "__torch__.torch.classes";
```
Comment (raziel) [inline torch/csrc/jit/mobile/import.cpp:425]: I'm a bit confused: if per your description the change is entirely FC, I assumed then that all the changes would take place in the serialization code.

However, if there's changes needed on import how can this feature be forward compatible?

**Review by raziel — COMMENTED** [review]
After looking at your code (and thinking about your description), I'm not sure the description saying the change is FC is correct. Otherwise why there's changes on import, and why you want a staged roll out?

To me, an FC change would imply that all changes take place in the serialization and consequently are entirely transparent to the runtime, so model after this change can execute in existing (i.e. older) runtimes.

**This is then an issue since the compatibility APIs would not work.**
The compatibility is currently defined in terms of the bytecode version, and the collection of operators contained in the model and supported by the runtime (i.e. the former being a subset of the latter).

However, currently, this new feature won't be tracked by neither: it's not triggering a version bump and it's not "an op".

File: torch/csrc/jit/mobile/import.cpp (hunk 2 above)
Comment (iseeyuan) [reply torch/csrc/jit/mobile/import.cpp:470]: A type is associated with the byte code of that function. For example, when running LIST_CONSTRUCT of a function, it fetches the type from the type table: `const auto& type = code_->types_[inst.X]->expectRef<at::ListType>();`

File: test/mobile/test_lite_script_type.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+from torch.jit.mobile import _load_for_lite_interpreter
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+
+class TestLiteScriptModule(TestCase):
+
+    def test_namedtuple_custom_classtype(self):
+        class Foo(NamedTuple):
+            id: torch.Tensor
```
Comment (iseeyuan) [inline test/mobile/test_lite_script_type.py:14]: Could you add more fields in this test, such as  `name: foo`, etc., for a more generic case?

Comment (iseeyuan) [reply test/mobile/test_lite_script_type.py:14]: How about just a NamedTuple, without a class wrapper? Could you also add it to the test?

File: test/mobile/test_lite_script_type.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        class Bar(torch.nn.Module):
+            def __init__(self):
+                super(Bar, self).__init__()
+                self.foo = Foo(torch.tensor(1))
```
Comment (iseeyuan) [inline test/mobile/test_lite_script_type.py:19]: It's great to have Class of NamedTuple supported in different scenarios: creation, return, attribute access. Would be nice to add those supported scenarios in the comments. 

Could you also verify if it has Function access?

**Comment by iseeyuan** [comment]
> After looking at your code (and thinking about your description), I'm not sure the description saying the change is FC is correct. Otherwise why there's changes on import, and why you want a staged roll out?
> 
> To me, an FC change would imply that all changes take place in the serialization and consequently are entirely transparent to the runtime, so model after this change can execute in existing (i.e. older) runtimes.
> 
> **This is then an issue since the compatibility APIs would not work.**
> The compatibility is currently defined in terms of the bytecode version, and the collection of operators contained in the model and supported by the runtime (i.e. the former being a subset of the latter).
> 
> However, currently, this new feature won't be tracked by neither: it's not triggering a version bump and it's not "an op".

@raziel , good points on FC! I think what @cccclai meant is that this change is not bound to a version bump, because if there's a model with a Class of NameTuple:
Before: it fails because the type is not supported.
After: if the new model is running by an old runtime, it still fails, because there's a new type that cannot be parsed. 
The failure reasons and behaviors are different, but either way, it does not make things "worse". 

But it's a great catch on compatibility API. It does not check types. So we could either
1. Bump the version so this change is checked in the API. Backporting would be exporting without this new type.
2. Add type check in the compatibility API.

**Comment by iseeyuan** [comment]
@cccclai , could you add a full program example (from torchscript to bytecode in the summary and comments), and provide a notebook with model inspector as @dhruvbird suggested? It would really help us understand the change.

**Comment by raziel** [comment]
> 1. Bump the version so this change is checked in the API. Backporting would be exporting without this new type.
> 2. Add type check in the compatibility API.

That's right.

A. We signal this change via the bytecode version.
B. We extend the compatibility API to check for a 3rd thing: whether a model contains specific custom classes, and what custom classes are supported by the runtime.

Option A:
* Pro: Has has the advantage of arguably simpler implementation: it's just a version bump, with no changes in the compatibility API.
* Con: It's a hammer for something that is more fine grained, and arguably is not even about versioning (as discussed above). Now all new models will be of a new version, when most wouldn't need it (i.e. do not use NamedTuple). Moreover, this is not even a full solution: it allows the compatibility API o work, but as you said, backporting would not be possible: the backporting function for models that do not use NamedTuple can just "do nothing", but models that do can only fail (since this is not a representation change, but a new feature not present before). On the flip side, we've never made the claim that backporting will always be successful/possible (side note: we should double check the workflow system accounts for it).

Option B:
* Pro: A more fine-grained compatibility check that will only fail for the most specific cases (a model that uses the feature and a runtime that doesn't). There's no need to version bump, nor backport.
* Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.

Let me know if I'm missing something with either option.
Overall, because of the issues I mention with using bytecode version for this I'm inclined to option B. However, I'm not sure how cleanly that can be implemented. If not, then option A would be best.

**Comment by iseeyuan** [comment]
> Let me know if I'm missing something with either option.
> Overall, because of the issues I mention with using bytecode version for this I'm inclined to option B. However, I'm not sure how cleanly that can be implemented. If not, then option A would be best.

My preference would be option B as well. I think we could maintain a supported type table in runtime, and compare it to the type table of the model. Type support check can be put in another PR(s). It could provide a systematic way of type compatibility. The check happens at earlier stage at delivery time, instead of at runtime. 

@cccclai , how do you think?

**Comment by JacobSzwejbka** [comment]
> * Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.

Do you think its necessary to have a representation more complex then just a list of the supported classes in this case clients would just return ["named tuple"] essentially and then we would do another partial deserialization of the model bytecode to find all its custom classes.

**Comment by cccclai** [comment]
> A. We signal this change via the bytecode version.
> B. We extend the compatibility API to check for a 3rd thing: whether a model contains specific custom classes, and what custom classes are supported by the runtime.

I think option B is probably better. Or I'd rephrase it to check if a model contains the type we support. If we need to support other types in the future, it's also easier to extend.

**Comment by raziel** [comment]
> > * Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.
> 
> Do you think its necessary to have a representation more complex then just a list of the supported classes in this case clients would just return ["named tuple"] essentially and then we would do another partial deserialization of the model bytecode to find all its custom classes.

Yeah, I don't think we need something complex, though I'd like to make sure that whatever it is contains all the information that we could need to disambiguate the types we're going to compare (and obviously that we could get that information from both the runtime API and from the model API).

What do you think?

**Comment by JacobSzwejbka** [comment]
> > > * Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.
> > 
> > 
> > Do you think its necessary to have a representation more complex then just a list of the supported classes in this case clients would just return ["named tuple"] essentially and then we would do another partial deserialization of the model bytecode to find all its custom classes.
> 
> Yeah, I don't think we need something complex, though I'd like to make sure that whatever it is contains all the information that we could need to disambiguate the types we're going to compare (and obviously that we could get that information from both the runtime API and from the model API).
> 
> What do you think?

Yeah that makes sense the implementation needs to be robust enough to be able to tell, im just not sure what other information we need beyond a name without a concrete example. I guess maybe at some point the idea is we only support certain uses of specific custom classes? Or custom classes themselves change and require essentially a versioning system of their own.

File: torch/csrc/jit/serialization/export_module.cpp [hunk 7]
```diff
@@ -190,9 +179,35 @@ std::pair<IValue, IValue> getFunctionTuple(
...
+        //   ('NamedTuple,
+        //     ('filed_name_1', 'field_type_1')
+        //     ('filed_name_2', 'field_type_2')
+        //   )
+        // )
+        IValue named_tuple_type_definition =
+            c10::ivalue::Tuple::create(std::vector<IValue>{
+                named_tuple, c10::ivalue::Tuple::create(name_type_pairs)});
+        types.emplace_back(c10::ivalue::Tuple::create(std::vector<IValue>{
+            IValue(t->str()), named_tuple_type_definition}));
```
Comment (pavithranrao) [inline torch/csrc/jit/serialization/export_module.cpp:207]: Do we need to use the mangled (unique) name of the type of the NamedTuple `__torch__. ...` by using  `type_name_uniquer_.getUniqueName` like we do for modules.

For reference please check this. Previously when we have sub modules that have same name, the names were not not mangled hence in `bytecode.pkl` there was only one reference.
https://github.com/pytorch/pytorch/pull/61933

as @iseeyuan mentioned, this might not be required for NamedTuple objects. Please correct if I am wrong, to check this, we have multiple instance of same NamedTuple object like,

```
self.foo_1 = Foo(torch.tensor(1))
self.foo_2 = Foo(torch.tensor(1))
...
```

to see if it gets saved and loaded as expected.

Comment (cccclai) [reply torch/csrc/jit/serialization/export_module.cpp:207]: Thanks for providing the example code. I check this case, and it looks fine. The source code I used is 
```
class Foo(NamedTuple):
    id: torch.Tensor

class Bar(torch.nn.Module):
    def __init__(self):
        super(Bar, self).__init__()
        # self.foo = Foo(torch.tensor(1))
        self.foo_1 = Foo(torch.tensor(1))
        self.foo_2 = Foo(torch.tensor(1))

    def forward(self, a: torch.Tensor):
        self.foo_1 = Foo(a)
        return (self.foo_1, self.foo_2)

```

and the corresponding TorchScript code is: 
```
class Bar(Module):
  __parameters__ = []
  __buffers__ = []
  training : bool
  _is_full_backward_hook : Optional[bool]
  foo_1 : Tuple[Tensor]
  foo_2 : Tuple[Tensor]
  def forward(self: __torch__.Bar,
    a: Tensor) -> Tuple[Tuple[Tensor], Tuple[Tensor]]:
    self.foo_1 = __torch__.Foo(a,)
    self.foo_2 = __torch__.Foo(a,)
    foo_1 = self.foo_1
    foo_2 = self.foo_2
    return (foo_1, foo_2)
class Foo(NamedTuple):
  id : Tensor
```
After thinking about it, each type name in the type table is associated with TorchScript source code (`__torch__.Foo` in the example case). This name should be unique, because class definition in the same python file under the same corresponding path is unique. For example, the following code will only have one valid `Foo` class definition in
`test.py`
```
class Foo(NamedTuple):
    id: torch.Tensor

class Foo(NamedTuple):
    id: torch.Tensor
    age: torch.Tensor
```

**Comment by dhruvbird** [comment]
> Yeah that makes sense the implementation needs to be robust enough to be able to tell, im just not sure what other information we need beyond a name without a concrete example. I guess maybe at some point the idea is we only support certain uses of specific custom classes? Or custom classes themselves change and require essentially a versioning system of their own.

Yeah, I'm worried that option-B will lead us to the path of requiring versioning the custom classes themselves - sort of boils down to a generic schema evolution check of a proto/thrift struct.

**Comment by dhruvbird** [comment]
> Before: it fails because the type is not supported.
> After: if the new model is running by an old runtime, it still fails, because there's a new type that cannot be parsed.
> The failure reasons and behaviors are different, but either way, it does not make things "worse".

The 2 failures happen at different times in the developer's workflow (I think).

Before: it isn't possible to export a model for mobile.

After: it's possible to export a model for mobile, but older runtimes can't load it if it contains a custom type that extends NameTuple

(I hope I understood that correctly).

**Comment by raziel** [comment]
> > Before: it fails because the type is not supported.
> > After: if the new model is running by an old runtime, it still fails, because there's a new type that cannot be parsed.
> > The failure reasons and behaviors are different, but either way, it does not make things "worse".
> 
> The 2 failures happen at different times in the developer's workflow (I think).
> 
> Before: it isn't possible to export a model for mobile.
> 
> After: it's possible to export a model for mobile, but older runtimes can't load it if it contains a custom type that extends NameTuple
> 
> (I hope I understood that correctly).


What @iseeyuan means is from the point of view of the runtime.
Whether the model was not possible to serialize and now is, it's just a function of adding a new capability. There's no going around that.

**Comment by iseeyuan** [comment]
> What @iseeyuan means is from the point of view of the runtime.
> Whether the model was not possible to serialize and now is, it's just a function of adding a new capability. There's no going around that.
That's right. My focus was more on FC of adding a new capability. 

@dhruvbird , 
>  it isn't possible to export a model for mobile.
It's correct for master code as is. It's a manual check we used at exporting stage to capture it earlier.

File: torch/csrc/jit/serialization/export_module.cpp (hunk 7 above)
Comment (pavithranrao) [reply torch/csrc/jit/serialization/export_module.cpp:207]: thanks for checking this and clarifying it 👍

**Comment by cccclai** [comment]
> > * Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.
> 
> Do you think its necessary to have a representation more complex then just a list of the supported classes in this case clients would just return ["named tuple"] essentially and then we would do another partial deserialization of the model bytecode to find all its custom classes.

It will be the same what JIT has been doing and we will need to 1) re-introduce the type-parser (possibily a few hundred k size regression) + 2) we need to parse the archive `code`, since it's the only place with source code.

**Comment by cccclai** [comment]
> What do you think?



> > > * Con: Likely more complex implementation (we would need a representation for "custom classes" that we can compare against). Looking forward, do we now need to think about versioning for custom classes as we do for ops? Probably not needed in practice, but theoretically possible.
> > 
> > 
> > Do you think its necessary to have a representation more complex then just a list of the supported classes in this case clients would just return ["named tuple"] essentially and then we would do another partial deserialization of the model bytecode to find all its custom classes.
> 
> Yeah, I don't think we need something complex, though I'd like to make sure that whatever it is contains all the information that we could need to disambiguate the types we're going to compare (and obviously that we could get that information from both the runtime API and from the model API).
> 
> What do you think?

I added the type checking api in #pr63129. Feel free to let me know what you think.

**Comment by cccclai** [comment]
> > Yeah that makes sense the implementation needs to be robust enough to be able to tell, im just not sure what other information we need beyond a name without a concrete example. I guess maybe at some point the idea is we only support certain uses of specific custom classes? Or custom classes themselves change and require essentially a versioning system of their own.
> 
> Yeah, I'm worried that option-B will lead us to the path of requiring versioning the custom classes themselves - sort of boils down to a generic schema evolution check of a proto/thrift struct.

I added the custom type parser in #pr63130 and we can check if the type from model is supported from the runtime. Feel free to let me know your thought.

File: test/mobile/test_lite_script_type.py (hunk 6 above)
Comment (iseeyuan) [reply test/mobile/test_lite_script_type.py:41]: Got it. So inlining actually helps. Maybe it's something to keep in mind that when we support CALL and not everything is inlined, we may revisit this issue.

**Review by iseeyuan — APPROVED** [review]
LGTM
